frontend/internal-packages/schema-bench/src/evaluate/evaluate.ts

```diff
@@ -0,0 +1,258 @@
+/**
+ * Database Schema Evaluation Script
+ *
+ * This script evaluates the accuracy of predicted database schemas against reference schemas.
+ * It performs comprehensive matching and scoring across multiple dimensions:
+ * - Table name mapping using word overlap and semantic similarity
+ * - Column name matching within each table
+ * - Primary key validation
+ * - Constraint validation
+ *
+ * The evaluation produces metrics including F1 scores, precision/recall, and all-correct rates
+ * to assess the quality of schema prediction models or tools.
+ */
+import type { PrimaryKeyConstraint, Schema } from '@liam-hq/db-structure'
+import { nameSimilarity } from '../nameSimilarity'
+import { wordOverlapMatch } from '../wordOverlapMatch'
+
+// Small epsilon value for numerical comparisons
+const EPSILON = 1e-5
+
+// Threshold for determining if all components are correct (table + column + primary key)
+// The value 2.9 represents the sum of perfect scores across three dimensions:
+// - Table name matching (1.0)
+// - Column name matching (1.0)
+// - Primary key validation (0.9)
+// This threshold ensures that all components must achieve near-perfect accuracy to be considered "all correct."
+const ALL_CORRECT_THRESHOLD = 2.9
+
+type Mapping = Record<string, string>
+
+type EvaluateResult = {
+  tableMapping: Mapping
+  columnMappings: Record<string, Mapping>
+  tableF1Score: number
+  tableAllCorrectRate: number
+  columnF1ScoreAverage: number
+  columnAllCorrectRateAverage: number
+  primaryKeyAccuracyAverage: number
+  constraintAccuracy: number
+  overallSchemaAccuracy: number
+}
+
+const createTableMapping = async (
+  referenceTableNames: string[],
+  predictTableNames: string[],
+): Promise<Mapping> => {
+  const tableMapping: Mapping = {}
+
+  // NOTE: Implement synonym matching if needed
+  // --- (0) synonym matching
+
+  // --- (1) name similarity matching
+  await nameSimilarity(referenceTableNames, predictTableNames, tableMapping)
+
+  // --- (2) word overlap matching
+  wordOverlapMatch(referenceTableNames, predictTableNames, tableMapping)
+
+  return tableMapping
+}
+
+const calculateTableMetrics = (
+  referenceTableNames: string[],
+  predictTableNames: string[],
+  tableMapping: Mapping,
+) => {
+  const tablePrecision =
+    predictTableNames.length === 0
+      ? 0
+      : Object.keys(tableMapping).length / predictTableNames.length
+  const tableRecall =
+    referenceTableNames.length === 0
+      ? 0
+      : Object.keys(tableMapping).length / referenceTableNames.length
+  const tableF1 =
+    tablePrecision + tableRecall === 0
+      ? 0
+      : (2 * tablePrecision * tableRecall) / (tablePrecision + tableRecall)
```

Comment:
This is the standard formula for computing the F1 score, which is the harmonic mean of precision and recall.

Reply:
A similar formula is also presented in the appendix of the paper we referenced during implementation

https://arxiv.org/html/2503.23886v1

> <img width="851" alt="2025-06-24 21 16 52" src="https://github.com/user-attachments/assets/2680c07d-54e3-489e-a9bc-96bd556b4c14" />